rqt_gui_py/src/rqt_gui_py/ros_py_plugin_provider.py

```diff
@@ -48,11 +48,13 @@ def __init__(self):
         self._node_initialized = False
         self._node = None
         self._spinner = None
+        self._shutdown_timeout = 2000
 
     def shutdown(self):
         qDebug('Shutting down RosPyPluginProvider')
         if self._spinner:
             self._spinner.quit()
+            self._spinner.wait(msecs=self._shutdown_timeout)
```

Comment:
If `wait()` timed out, cleaning up the node will still be an issue.
Maybe something like:
```suggestion
            joined = self._spinner.wait(msecs=self._shutdown_timeout)
        if not joined:
            # maybe log an error
            return
```

The other option is to not pass a timeout, and block forever until joining the thread.

Reply:
No option is ideal:

* If there is a return after the the wait and the wait times out, the app will exit, but the node won't get cleaned up and will still appear when doing "ros2 node list." 
* If the wait times out and there is no return statement (the code continues, cleaning up the node), there is still a risk of a crash (although slight). If there is a crash, the node is not cleaned up properly and still appears when doing "ros2 node list"
* If waiting without a timeout, if there is an issue with the spinner thread, the app will hang

I suppose from the user's perspective, the first option would be the best. At least there wouldn't be a crash or hang. The node would still appear in the list for a while. What do you think?

Reply:
> I suppose from the user's perspective, the first option would be the best. At least there wouldn't be a crash or hang. The node would still appear in the list for a while. What do you think?

:+1: the first option sounds reasonble to me.
The node will dissappear after a time of `ros2 node list` too, so that's not a big issue.